packages/financial-templates-lib/src/price-feed/USPACPriceFeed.ts

```diff
@@ -0,0 +1,243 @@
+    const url = `https://stock-data-yahoo-finance-alternative.p.rapidapi.com/v8/finance/spark?symbols=${symbolsStr}&range=${range}&interval=${this.interval}`;
+    const options = {
+      method: "GET",
+      headers: {
+        "x-rapidapi-host": "stock-data-yahoo-finance-alternative.p.rapidapi.com",
+        "x-rapidapi-key": this.rapidApiKey,
+      },
+    };
+    const response = (await this.networker.getJson(url, options)) as RapidAPIHistoryReponse | null | undefined;
+
+    if (response == null) {
+      throw new Error(`🚨Invalid response from url ${url}`);
+    }
+
+    if (verbose) {
+      console.log(`Rapid API response for url ${url} is ${JSON.stringify(response, null, 2)}`);
+    }
+
+    // Validate response
+    for (const symbol of this.symbols) {
+      const stockData = response[symbol];
+      if (stockData == null) {
+        throw new Error(`🚨Could not parse price result from url ${url}: missing data for symbol ${symbol}`);
+      }
+      if (stockData.timestamp.length == 0 || stockData.timestamp.length != stockData.close.length) {
+        throw new Error(`🚨Could not parse price result from url ${url}: invalid data for symbol ${symbol}`);
+      }
+      for (const ts of stockData.timestamp) {
+        if (typeof ts != "number") {
+          throw new Error("Could not parse data: invalid timestamp type");
+        }
+      }
+      for (const p of stockData.close) {
+        if (typeof p != "number") {
+          throw new Error("Could not parse data: invalid price type");
+        }
+      }
+      for (let i = 1; i < stockData.timestamp.length; i++) {
+        if (stockData.timestamp[i] <= stockData.timestamp[i - 1]) {
+          throw new Error(`Could not parse data: timestamps for ${symbol} are not increasing monotonously`);
+        }
+      }
+    }
+
+    // Get prices
+    const prices = this.symbols.map((symbol) => {
+      const stockData = response[symbol];
+
+      // iterate over timestamps and prices backwards, until we find timestamp
+      // that is earlier then given time
+      const result = lodash
+        .zip(stockData.timestamp, stockData.close)
+        .reverse()
+        .find((item) => time >= (item[0] as number));
+      if (result == null) {
+        throw new Error(`Could not get historical price for symbol ${symbol}: timestamp ${time} past last data point`);
+      }
+      const price = result[1] as number;
+      return price;
+    });
+
+    return this.calculateBasketPrice(prices);
+  }
+
+  public getLastUpdateTime(): number | null {
+    return this.lastUpdateTime;
+  }
+
+  public getLookback(): number {
+    return this.lookback;
+  }
+
+  public getPriceFeedDecimals(): number {
+    return this.priceFeedDecimals;
+  }
+
+  public async update(): Promise<void> {
+    const currentTime = await this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== null && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
```

Comment:
Is it correct to assume that this check will only work in a set-up that retains state between runs? For instance, there are bots that run in a serverless mode and will not know when the last update time was. Would the consequence be that those bots would exceed the limit on the free plan for the API? (@Reinis-FRP does that sound correct to you?) Any other consequences I'm not considering?

Reply:
Yes, in order to not run out of free tier limits the bot should be run in environment capable of keeping the state (i.e. using POLLING_DELAY > 0 or customizing the bot to save lastUpdateTime to some other medium when running in serverless mode).